open_seq2seq/decoders/convs2s_decoder.py

```diff
@@ -0,0 +1,318 @@
+import math
+from .decoder import Decoder
+
+from open_seq2seq.parts.transformer import beam_search
+
+from open_seq2seq.parts.transformer import embedding_layer
+from open_seq2seq.parts.transformer.utils import get_padding
+
+from open_seq2seq.parts.convs2s import ffn_wn_layer, conv_wn_layer, attention_wn_layer
+
+# Default value used if max_input_length is not given
+MAX_INPUT_LENGTH = 128
+
+
+class ConvS2SDecoder(Decoder):
+  @staticmethod
+  def get_required_params():
+    """Static method with description of required parameters.
+
+      Returns:
+        dict:
+            Dictionary containing all the parameters that **have to** be
+            included into the ``params`` parameter of the
+            class :meth:`__init__` method.
+    """
+    return dict(Decoder.get_required_params(), **{
+      'batch_size': int,
+      'decoder_layers': int,
+
+      'tgt_emb_size': int,
+      'tgt_vocab_size': int,
+      'shared_embed': bool,
+      'embedding_dropout_keep_prob': float,
+
+      'hidden_dropout_keep_prob': float,
+      'out_dropout_keep_prob': float,
+
+      'beam_size': int,
+      'alpha': float,
+      'extra_decode_length': int,
+
+      'EOS_ID': int,
+
+    })
+
+  @staticmethod
+  def get_optional_params():
+    """Static method with description of optional parameters.
+
+      Returns:
+        dict:
+            Dictionary containing all the parameters that **can** be
+            included into the ``params`` parameter of the
+            class :meth:`__init__` method.
+    """
+    return dict(Decoder.get_optional_params(), **{
+      'pad_embeddings_2_eight': bool,
+      'conv_knum': list,
+      'conv_kwidth': list,
+
+      # if not provided, tgt_emb_size is used as the default value
+      'out_emb_size': int,
+
+      'max_input_length': int,
+      'GO_SYMBOL': int,
+      'PAD_SYMBOL': int,
+      'END_SYMBOL': int,
+    })
+
+  def _cast_types(self, input_dict):
+    return input_dict
+
+  def __init__(self, params, model, name="convs2s_decoder", mode='train'):
+    super(ConvS2SDecoder, self).__init__(params, model, name, mode)
+    self.embedding_softmax_layer = None
+    self.position_embedding_layer = None
+    self.layers = []
+    self._tgt_vocab_size = self.params['tgt_vocab_size']
+    self._tgt_emb_size = self.params['tgt_emb_size']
+    self._mode = mode
+    self._pad_sym = self.params.get('PAD_SYMBOL', 0)
+    self._pad2eight = params.get('pad_embeddings_2_eight', False)
+
+
+  def _decode(self, input_dict):
+    targets = input_dict['target_tensors'][0] if 'target_tensors' in input_dict else None
+
+    encoder_outputs = input_dict['encoder_output']['outputs']
+    encoder_outputs_b = input_dict['encoder_output'].get('outputs_b', encoder_outputs)
+
+    inputs_attention_bias = input_dict['encoder_output'].get('inputs_attention_bias_cs2s', None)
+
+    with tf.name_scope("decode"):
+      # prepare decoder layers
+      if len(self.layers) == 0:
+        knum_list = self.params.get("conv_knum")
```

Comment:
.get will append None as the default if default explicitly not specified. Did you mean to have self.params.get("conv_knum", []) here? Or not use .get and move this parameter to required_params?